src/bookshelf/schema/deploy/base_schema.sql

```diff
@@ -0,0 +1,199 @@
+
+CREATE OR REPLACE VIEW expanded_files AS SELECT b.name AS bucket_name, fd.*
+     FROM buckets b
+     INNER JOIN
+       (SELECT bucket_id, name, created_at, file_id, d.* FROM
+               file_names f INNER JOIN file_data d ON f.data_id = d.data_id) fd
+	ON b.id = fd.bucket_id;
+
+--
+-- This atomically links a file data item with
+--
+CREATE OR REPLACE FUNCTION link_file_data (
+       my_file_id file_names.file_id%TYPE,
+       my_data_id file_data.data_id%TYPE
+       )
+       RETURNS VOID -- fix this to be null
+AS $$
+BEGIN
+  UPDATE file_data SET complete = true WHERE data_id = my_data_id;
+  UPDATE file_names SET data_id = my_data_id WHERE file_id = my_file_id;
+END
+$$
+LANGUAGE plpgsql VOLATILE;
+
+--
+-- This could be refactored to use link_file_data above, but it seems
+-- simpler to do it this way.
+--
+CREATE OR REPLACE FUNCTION create_file_link_data(
+       bn buckets.name%TYPE,
+       my_name  file_names.name%TYPE,
+       my_data_id file_data.data_id%TYPE)
+RETURNS file_names.file_id%TYPE -- what happens if exists already? UPSERT?
+AS $$
+DECLARE
+   my_bucket_id buckets.id%TYPE;
+   new_file_id file_names.file_id%TYPE;
+BEGIN
+   SELECT b.id INTO my_bucket_id FROM buckets AS b WHERE b.name = bn;
+   UPDATE file_data SET complete = true WHERE data_id = my_data_id;
+   INSERT INTO file_names (bucket_id, "name", data_id) VALUES (my_bucket_id, my_name, my_data_id)
+   	       RETURNING file_id INTO new_file_id;
+   RETURN new_file_id;
+END;
+$$
+LANGUAGE plpgsql VOLATILE;
+
+CREATE OR REPLACE FUNCTION mark_deleted(id file_data.data_id%TYPE)
+RETURNS file_data.tombstoned_at%TYPE
+AS $$
+DECLARE
+   my_tombstone file_data.tombstoned_at%TYPE;
+BEGIN
+   SELECT (now() at time zone 'utc') INTO my_tombstone;
+   UPDATE file_data SET tombstoned_at = my_tombstone WHERE data_id = id AND tombstoned_at IS NULL;
+   RETURN my_tombstone;
+END;
+$$
+LANGUAGE plpgsql VOLATILE;
+
+CREATE OR REPLACE FUNCTION delete_file_last_reference() RETURNS TRIGGER as $delete_file_last_reference$
+   BEGIN
+       IF NOT EXISTS (SELECT 1 FROM file_names WHERE data_id = OLD.data_id) THEN
+         PERFORM mark_deleted(OLD.data_id);
+       END IF;
+       RETURN NULL;
+   END;
+$delete_file_last_reference$ LANGUAGE plpgsql;
+
+CREATE TRIGGER delete_file_last_reference_on_delete AFTER DELETE ON file_names FOR EACH ROW EXECUTE PROCEDURE delete_file_last_reference();
+CREATE TRIGGER delete_file_last_reference_on_update AFTER UPDATE ON file_names FOR EACH ROW EXECUTE PROCEDURE delete_file_last_reference();
+
+-- These functions are called by bookshelf regularly
+CREATE OR REPLACE FUNCTION purge_expired(delay integer)
+RETURNS integer
+AS $$
+DECLARE
+   expire_time file_data.tombstoned_at%TYPE;
+   deleted integer;
+BEGIN
+   SELECT (now() at time zone 'utc') - (delay * '1 millisecond'::interval) INTO expire_time;
+   WITH a AS (DELETE FROM file_data WHERE tombstoned_at < expire_time RETURNING 1)
+   SELECT count(*) FROM a INTO deleted;
+   RETURN deleted;
```

Comment:
Some comments lost their place on rebase.  Previously @sdelano asked:

> What's the difference here between just returning the results of SELECT instead of doing a SELECT INTO followed by a RETURN?

And @marcparadise followed up with:

> Is the intent here just to indicate if something was deleted or not? If so, can we just use a normal delete query from sqerl, and check if the returned number of affected rows is > 1?

Same for cleanup_abandoned_uploads, can it be justa straight up delete instead?